<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <link rel="stylesheet" href="/seqfu2/assets/css/just-the-docs-default.css"> <link rel="stylesheet" href="/seqfu2/assets/css/just-the-docs-head-nav.css" id="jtd-head-nav-stylesheet"> <style id="jtd-nav-activation"> .site-nav > ul.nav-list:first-child > li > a, .site-nav > ul.nav-list:first-child > li > ul > li:not(:nth-child(18)) > a, .site-nav > ul.nav-list:first-child > li > ul > li > ul > li a { background-image: none; } .site-nav > ul.nav-list:not(:first-child) a, .site-nav li.external a { background-image: none; } .site-nav > ul.nav-list:first-child > li:nth-child(5) > ul > li:nth-child(18) > a { font-weight: 600; text-decoration: none; } .site-nav > ul.nav-category-list > li > button svg, .site-nav > ul.nav-list:first-child > li:nth-child(5) > button svg, .site-nav > ul.nav-list:first-child > li:nth-child(5) > ul > li:nth-child(18) > button svg { transform: rotate(-90deg); } .site-nav > ul.nav-category-list > li.nav-list-item > ul.nav-list, .site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(5) > ul.nav-list, .site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(5) > ul.nav-list > li.nav-list-item:nth-child(18) > ul.nav-list { display: block; } </style> <script src="/seqfu2/assets/js/vendor/lunr.min.js"></script> <script src="/seqfu2/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.8.0 --> <title>seqfu stats | SeqFu Documentation</title> <meta name="generator" content="Jekyll v3.10.0" /> <meta property="og:title" content="seqfu stats" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="A general-purpose toolkit to manipulate and parse FASTA/FASTQ files" /> <meta property="og:description" content="A general-purpose toolkit to manipulate and parse FASTA/FASTQ files" /> <link rel="canonical" href="http://0.0.0.0:4000/seqfu2/tools/stats.html" /> <meta property="og:url" content="http://0.0.0.0:4000/seqfu2/tools/stats.html" /> <meta property="og:site_name" content="SeqFu Documentation" /> <meta property="og:type" content="website" /> <meta name="twitter:card" content="summary" /> <meta property="twitter:title" content="seqfu stats" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"A general-purpose toolkit to manipulate and parse FASTA/FASTQ files","headline":"seqfu stats","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"http://0.0.0.0:4000/seqfu2/img/seqfu-512.png"}},"url":"http://0.0.0.0:4000/seqfu2/tools/stats.html"}</script> <!-- End Jekyll SEO tag --> </head> <body> <a class="skip-to-main" href="#main-content">Skip to main content</a> <svg xmlns="http://www.w3.org/2000/svg" class="d-none"> <symbol id="svg-link" viewBox="0 0 24 24"> <title>Link</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"> <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"> <title>Menu</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"> <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"> <title>Expand</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"> <polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE --> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"> <title id="svg-external-link-title">(external link)</title> <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"> <title>Document</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"> <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"> <title>Search</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <!-- Bootstrap Icons. MIT License: https://github.com/twbs/icons/blob/main/LICENSE.md --> <symbol id="svg-copy" viewBox="0 0 16 16"> <title>Copy</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16"> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"/> <path d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"/> </svg> </symbol> <symbol id="svg-copied" viewBox="0 0 16 16"> <title>Copied</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard-check-fill" viewBox="0 0 16 16"> <path d="M6.5 0A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3Zm3 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3Z"/> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1A2.5 2.5 0 0 1 9.5 5h-3A2.5 2.5 0 0 1 4 2.5v-1Zm6.854 7.354-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 0 1 .708-.708L7.5 10.793l2.646-2.647a.5.5 0 0 1 .708.708Z"/> </svg> </symbol> </svg> <div class="side-bar"> <div class="site-header" role="banner"> <a href="/seqfu2/" class="site-title lh-tight"> <div class="site-logo" role="img" aria-label="SeqFu Documentation"></div> </a> <button id="menu-button" class="site-button btn-reset" aria-label="Toggle menu" aria-pressed="false"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-menu"></use></svg> </button> </div> <nav aria-label="Main" id="site-nav" class="site-nav"> <ul class="nav-list"><li class="nav-list-item"><a href="/seqfu2/readme.html" class="nav-list-link">Home</a></li><li class="nav-list-item"><a href="/seqfu2/" class="nav-list-link">Home</a></li><li class="nav-list-item"><a href="/seqfu2/installation" class="nav-list-link">Installation</a></li><li class="nav-list-item"><a href="/seqfu2/intro" class="nav-list-link">Overview</a></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Core Tools category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/seqfu2/tools/README.html" class="nav-list-link">Core Tools</a><ul class="nav-list"><li class="nav-list-item"><a href="/seqfu2/tools/bases.html" class="nav-list-link">seqfu bases</a></li><li class="nav-list-item"><a href="/seqfu2/tools/cat.html" class="nav-list-link">seqfu cat</a></li><li class="nav-list-item"><a href="/seqfu2/tools/check.html" class="nav-list-link">seqfu check</a></li><li class="nav-list-item"><a href="/seqfu2/tools/count.html" class="nav-list-link">seqfu count</a></li><li class="nav-list-item"><a href="/seqfu2/tools/deinterleave.html" class="nav-list-link">seqfu deinterleave</a></li><li class="nav-list-item"><a href="/seqfu2/tools/derep.html" class="nav-list-link">seqfu derep</a></li><li class="nav-list-item"><a href="/seqfu2/tools/grep.html" class="nav-list-link">seqfu grep</a></li><li class="nav-list-item"><a href="/seqfu2/tools/head.html" class="nav-list-link">seqfu head</a></li><li class="nav-list-item"><a href="/seqfu2/tools/interleave.html" class="nav-list-link">seqfu interleave</a></li><li class="nav-list-item"><a href="/seqfu2/tools/lanes.html" class="nav-list-link">seqfu lanes</a></li><li class="nav-list-item"><a href="/seqfu2/tools/list.html" class="nav-list-link">seqfu list</a></li><li class="nav-list-item"><a href="/seqfu2/tools/merge.html" class="nav-list-link">seqfu merge</a></li><li class="nav-list-item"><a href="/seqfu2/tools/metadata.html" class="nav-list-link">seqfu metadata</a></li><li class="nav-list-item"><a href="/seqfu2/tools/qual.html" class="nav-list-link">seqfu qual</a></li><li class="nav-list-item"><a href="/seqfu2/tools/rc.html" class="nav-list-link">seqfu rc</a></li><li class="nav-list-item"><a href="/seqfu2/tools/rotate.html" class="nav-list-link">seqfu rotate</a></li><li class="nav-list-item"><a href="/seqfu2/tools/sort.html" class="nav-list-link">seqfu sort</a></li><li class="nav-list-item"><a href="/seqfu2/tools/stats.html" class="nav-list-link">seqfu stats</a></li><li class="nav-list-item"><a href="/seqfu2/tools/tabulate.html" class="nav-list-link">seqfu tabulate</a></li><li class="nav-list-item"><a href="/seqfu2/tools/tail.html" class="nav-list-link">seqfu tail</a></li><li class="nav-list-item"><a href="/seqfu2/tools/view.html" class="nav-list-link">seqfu view</a></li></ul></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Utilities category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/seqfu2/utilities/README.html" class="nav-list-link">Utilities</a><ul class="nav-list"><li class="nav-list-item"><a href="/seqfu2/utilities/fu-16Sregion.html" class="nav-list-link">fu-16Sregion</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-cov.html" class="nav-list-link">fu-cov</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-homocomp.html" class="nav-list-link">fu-homocom</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-index.html" class="nav-list-link">fu-index</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-msa.html" class="nav-list-link">fu-msa</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-multirelabel.html" class="nav-list-link">fu-multirelabel</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-nanotags.html" class="nav-list-link">fu-nanotags</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-orf.html" class="nav-list-link">fu-orf</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-primers.html" class="nav-list-link">fu-primers</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-shred.html" class="nav-list-link">fu-shred</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-sw.html" class="nav-list-link">fu-sw</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-tabcheck.html" class="nav-list-link">fu-tabcheck</a></li><li class="nav-list-item"><a href="/seqfu2/utilities/fu-virfilter.html" class="nav-list-link">fu-virfilter</a></li></ul></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Helper Utilities category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/seqfu2/scripts/README.html" class="nav-list-link">Helper Utilities</a><ul class="nav-list"><li class="nav-list-item"><a href="/seqfu2/scripts/fu-pecheck.html" class="nav-list-link">fu-pecheck</a></li><li class="nav-list-item"><a href="/seqfu2/scripts/fu-split.html" class="nav-list-link">fu-split</a></li></ul></li><li class="nav-list-item"><a href="/seqfu2/usage" class="nav-list-link">Usage Guide</a></li><li class="nav-list-item"><a href="/seqfu2/about" class="nav-list-link">About</a></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Releases category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/seqfu2/releases/README.html" class="nav-list-link">Releases</a><ul class="nav-list"><li class="nav-list-item"><a href="/seqfu2/releases/history.html" class="nav-list-link">version 1.3.6</a></li></ul></li><li class="nav-list-item"><a href="/seqfu2/tools/trim.html" class="nav-list-link">trim</a></li></ul> <div class="nav-category">Core Tools</div> <ul class="nav-list"></ul> <div class="nav-category">Utilities</div> <ul class="nav-list"></ul> <div class="nav-category">Helper Utilities</div> <ul class="nav-list"></ul> </nav> <footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll. </footer> </div> <div class="main" id="top"> <div id="main-header" class="main-header"> <div class="search" role="search"> <div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search SeqFu Documentation" aria-label="Search SeqFu Documentation" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label> </div> <div id="search-results" class="search-results"></div> </div> <nav aria-label="Auxiliary" class="aux-nav"> <ul class="aux-nav-list"> <li class="aux-nav-list-item"> <a href="https://github.com/telatin/seqfu2" class="site-button" target="_blank" rel="noopener noreferrer" > GitHub Repository </a> </li> <li class="aux-nav-list-item"> <a href="https://github.com/telatin/seqfu2/issues" class="site-button" target="_blank" rel="noopener noreferrer" > Report Issue </a> </li> </ul> </nav> </div> <div class="main-content-wrap"> <nav aria-label="Breadcrumb" class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="/seqfu2/tools/README.html">Core Tools</a></li> <li class="breadcrumb-nav-list-item"><span>seqfu stats</span></li> </ol> </nav> <div id="main-content" class="main-content"> <main> <h1 id="seqfu-stats"> <a href="#seqfu-stats" class="anchor-heading" aria-labelledby="seqfu-stats"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> seqfu stats </h1> <p><em>stats</em> is one of the core subprograms of <em>SeqFu</em>.</p><pre><code class="language-note">Version prior to v1.22.1 were calculatin auN incorrectly
</code></pre><div class="language-text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Usage: stats [options] [&lt;inputfile&gt; ...]

Options:
  -a, --abs-path         Print absolute paths
  -b, --basename         Print only filenames
  -n, --nice             Print nice terminal table
  -j, --json             Print json (EXPERIMENTAL)
  -s, --sort-by KEY      Sort by KEY from: filename, counts, n50, tot, avg, min, max
                         descending for values, ascending for filenames [default: none]
  -r, --reverse          Reverse sort order
  -t, --thousands        Add thousands separator (only tabbed/nice output)
  --csv                  Separate output by commas instead of tabs
  --gc                   Also print %GC
  --multiqc FILE         Saves a MultiQC report to FILE (suggested: name_mqc.txt)
  --precision INT        Number of decimal places to round to [default: 2]
  --noheader             Do not print header
  -v, --verbose          Verbose output
  -h, --help             Show this help
</code></pre></div></div> <h3 id="sorting"> <a href="#sorting" class="anchor-heading" aria-labelledby="sorting"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Sorting </h3><pre><code class="language-note">Sorting added in SeqFu 1.11. 
</code></pre><p>To sort by filename (ascending alphabetical order) add <code class="language-plaintext highlighter-rouge">--sort filename</code>. Numerical values are sorted from the largest (descending), supported values are:</p> <ul> <li><em>n50</em>, <em>n75</em> or <em>N90</em></li> <li><em>count</em> or <em>counts</em> (number of reads)</li> <li><em>sum</em> or <em>tot</em> (total bases)</li> <li><em>min</em> or <em>minimum</em> (minimum length)</li> <li><em>max</em> or <em>maximum</em> (maximum length)</li> <li><em>avg</em> or <em>mean</em> (average length)</li> <li><em>aun</em> (area under the Nx curve)</li> </ul> <p><strong>NOTE</strong> Specifying an invalid key will result in unsorted results with a warning, but in future releases this might throw an error.</p> <h3 id="example-output"> <a href="#example-output" class="anchor-heading" aria-labelledby="example-output"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Example output </h3> <p>Output is a TSV text with three columns (or CSV using <code class="language-plaintext highlighter-rouge">--csv</code>):</p> <div class="language-text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>File,#Seq,Sum,Avg,N50,N75,N90,Min,Max
data/filt.fa.gz,78730,24299931,308.6,316,316,220,180,485
</code></pre></div></div> <h3 id="screen-friendly-output"> <a href="#screen-friendly-output" class="anchor-heading" aria-labelledby="screen-friendly-output"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Screen friendly output </h3> <p>When using <code class="language-plaintext highlighter-rouge">-n</code> (<code class="language-plaintext highlighter-rouge">--nice</code>) output:</p> <div class="language-text highlighter-rouge"><div class="highlight"><pre class="highlight"><code>seqfu stats data/filt.fa.gz  -n
┌─────────────────┬───────┬──────────┬───────┬─────┬─────┬─────┬───────┬─────┬─────┐
│ File            │ #Seq  │ Total bp │ Avg   │ N50 │ N75 │ N90 │ auN   │ Min │ Max │
├─────────────────┼───────┼──────────┼───────┼─────┼─────┼─────┼───────┼─────┼─────┤
│ data/filt.fa.gz │ 78730 │ 24299931 │ 308.6 │ 316 │ 316 │ 220 │ 0.385 │ 180 │ 485 │
└─────────────────┴───────┴──────────┴───────┴─────┴─────┴─────┴───────┴─────┴─────┘
</code></pre></div></div> <h2 id="multiqc-output"> <a href="#multiqc-output" class="anchor-heading" aria-labelledby="multiqc-output"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> MultiQC output </h2> <p>Using the <code class="language-plaintext highlighter-rouge">--multiqc OUTPUTFILE</code> option it’s possible to save a MultiQC compatible file (we recommend to use the <em>projectname_mqc.tsv</em> filename format). After coolecting all the MultiQC files in a directory, using <code class="language-plaintext highlighter-rouge">multiqc -f .</code> will generate the MultiQC report. MultiQC itself can be installed via Bioconda with <code class="language-plaintext highlighter-rouge">conda install -y -c bioconda multiqc</code>.</p> <p>To understand how to use MultiQC, if you never did so, check their excellent <a href="https://multiqc.info">documentation</a>.</p> <h2 id="legacy"> <a href="#legacy" class="anchor-heading" aria-labelledby="legacy"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Legacy </h2> <p>The pre 1.11 version of the statistics has been made available via <code class="language-plaintext highlighter-rouge">seqfu oldstats</code>. There are no breaking changes at the moment, and an expanded set of tests ensures the compatibility not only of the metrics (unchanged) but also of the output (now supporting sorting options).</p> <h2 id="benchmark"> <a href="#benchmark" class="anchor-heading" aria-labelledby="benchmark"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Benchmark </h2> <p>A similar functionality is provided by <code class="language-plaintext highlighter-rouge">SeqKit</code>, so we compared the performance of SeqFu with <a href="https://bioinf.shenwei.me/seqkit/">SeqKit</a> and <a href="https://metacpan.org/pod/release/PROCH/Proch-N50-1.3.0/bin/n50">n50</a>, both available from bioconda. We used a Linux Virtual Machine running Ubuntu 18.04, with 8 cores and 64 Gb of RAM for the test, with Miniconda (4.9.2) to install the required tools.</p> <p>:warning: SeqKit, by default, omits N50 calculation, that is a core feature (always enabled) in SeqFu. The correct comparison is thus between <code class="language-plaintext highlighter-rouge">seqfu stats</code> and <code class="language-plaintext highlighter-rouge">seqkit stats --all</code>.</p> <p>Speed evaluate with <a href="https://github.com/sharkdp/hyperfine">hyperfine</a>, peak memory usage with this <a href="https://gist.github.com/MattForshaw/86b82b6c09bdbfce5ff5ee570e8a8bef">bash script</a>.</p> <p>As dataset we used the Human Genome (see <a href="https://bioinf.shenwei.me/seqkit/benchmark/">this benchmark page</a>), which contains few large sequences, and the reference genome of the gastrointestinal tract, which is composed by many short sequences instead.</p> <p>The test can be replicated with these commands:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Download tools (can be done in a new environment)</span>
conda <span class="nb">install</span> <span class="nt">-c</span> conda-forge <span class="nt">-c</span> bioconda hyperfine <span class="nv">n50</span><span class="o">=</span>1.3.0 <span class="nv">seqkit</span><span class="o">=</span>0.16.0 <span class="nv">seqfu</span><span class="o">=</span>0.9.6
 
<span class="c"># Download datasets: many short sequences and few large sequences </span>
wget http://downloads.hmpdacc.org/data/reference_genomes/body_sites/Gastrointestinal_tract.nuc.fsa
wget ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz

<span class="c"># Compare execution times</span>
<span class="nv">FILE1</span><span class="o">=</span>Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz
<span class="nv">FILE2</span><span class="o">=</span>Gastrointestinal_tract.nuc.fsa

<span class="k">for </span>FILE <span class="k">in</span> <span class="nv">$FILE1</span> <span class="nv">$FILE2</span><span class="p">;</span> <span class="k">do
  </span>hyperfine <span class="nt">--export-markdown</span> stat_<span class="si">$(</span><span class="nb">basename</span> <span class="nv">$FILE</span> | <span class="nb">cut</span> <span class="nt">-f1</span> <span class="nt">-d</span>.<span class="si">)</span>.md <span class="nt">--warmup</span> 1 <span class="nt">--min-runs</span> 3 <span class="se">\</span>
    <span class="s2">"seqfu stats </span><span class="nv">$FILE</span><span class="s2">"</span> <span class="se">\</span>
    <span class="s2">"seqkit stats </span><span class="nv">$FILE</span><span class="s2">"</span> <span class="se">\</span>
    <span class="s2">"seqkit stats --all </span><span class="nv">$FILE</span><span class="s2">"</span> <span class="se">\</span>
    <span class="s2">"n50 -x </span><span class="nv">$FILE</span><span class="s2">"</span> 
<span class="k">done</span>
</code></pre></div></div> <p>The result, for the <strong>Human Genome</strong> (few large sequences), has been:</p> <div class="table-wrapper"><table> <thead> <tr> <th style="text-align: left">Command</th> <th style="text-align: right">Mean [s]</th> <th style="text-align: right">Min [s]</th> <th style="text-align: right">Max [s]</th> <th style="text-align: right">Relative</th> </tr> </thead> <tbody> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">seqfu stats $FILE</code></td> <td style="text-align: right">27.420 ± 0.349</td> <td style="text-align: right">27.090</td> <td style="text-align: right">27.785</td> <td style="text-align: right">1.00</td> </tr> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">seqkit stats $FILE</code> :warning:</td> <td style="text-align: right">116.693 ± 0.236</td> <td style="text-align: right">116.512</td> <td style="text-align: right">116.960</td> <td style="text-align: right">4.26 ± 0.05</td> </tr> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">seqkit stats --all $FILE</code></td> <td style="text-align: right">120.435 ± 0.434</td> <td style="text-align: right">120.054</td> <td style="text-align: right">120.907</td> <td style="text-align: right">4.39 ± 0.06</td> </tr> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">n50 -x $FILE</code></td> <td style="text-align: right">34.888 ± 0.628</td> <td style="text-align: right">34.167</td> <td style="text-align: right">35.316</td> <td style="text-align: right">1.27 ± 0.03</td> </tr> </tbody> </table></div> <p>For the <strong>Gastrointestinal reference genomes</strong> (many short sequences):</p> <div class="table-wrapper"><table> <thead> <tr> <th style="text-align: left">Command</th> <th style="text-align: right">Mean [s]</th> <th style="text-align: right">Min [s]</th> <th style="text-align: right">Max [s]</th> <th style="text-align: right">Relative</th> </tr> </thead> <tbody> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">seqfu stats $FILE</code></td> <td style="text-align: right">6.885 ± 0.307</td> <td style="text-align: right">6.602</td> <td style="text-align: right">7.211</td> <td style="text-align: right">1.82 ± 0.09</td> </tr> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">seqkit stats $FILE</code> :warning:</td> <td style="text-align: right">3.793 ± 0.082</td> <td style="text-align: right">3.699</td> <td style="text-align: right">3.854</td> <td style="text-align: right">1.00</td> </tr> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">seqkit stats --all $FILE</code></td> <td style="text-align: right">7.667 ± 0.081</td> <td style="text-align: right">7.583</td> <td style="text-align: right">7.746</td> <td style="text-align: right">2.02 ± 0.05</td> </tr> <tr> <td style="text-align: left"><code class="language-plaintext highlighter-rouge">n50 -x $FILE</code></td> <td style="text-align: right">76.377 ± 1.990</td> <td style="text-align: right">74.891</td> <td style="text-align: right">78.638</td> <td style="text-align: right">20.14 ± 0.68</td> </tr> </tbody> </table></div> <h2 id="screenshot"> <a href="#screenshot" class="anchor-heading" aria-labelledby="screenshot"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Screenshot </h2> <p><img src="/seqfu2/img/screenshot-stats.svg" alt="Screenshot of &quot;seqfu stats&quot;" title="SeqFu stats" /></p> </main> <hr> <footer> <p><a href="#top" id="back-to-top">Back to top</a></p> <p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2019-2025 Andrea Telatin. Distributed by an <a href="https://github.com/telatin/seqfu2/blob/main/LICENSE">MIT license</a>.</p> <div class="d-flex mt-2"> </div> </footer> </div> </div> <button id="search-button" class="search-button btn-reset" aria-label="Focus on search"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-search"></use></svg> </button> <div class="search-overlay"></div> </div> </body> </html>
